shuttle/src/runtime/execution.rs

```diff
@@ -89,159 +151,176 @@ impl Execution {
                 caller,
             );
 
-            // Run the test to completion
-            while self.step(config) {}
+                // Run the test to completion
+                match self.run_to_competion(config.immediately_return_on_panic) {
+                    Ok(()) => {},
+                    Err(e) => {
+                        match e {
```

Comment:
Could define a `should_persist_failure` on `StepError` and just call

`if e.should_persist_failure() { persist_failure(config); }`

or even `e.persist_failure(config)`

before the match, instead of calling it in each branch